src/slice.rs

```diff
@@ -282,188 +303,323 @@ impl_sliceorindex_from_index!(isize);
 impl_sliceorindex_from_index!(usize);
 impl_sliceorindex_from_index!(i32);
 
+impl From<NewAxis> for AxisSliceInfo {
+    #[inline]
+    fn from(_: NewAxis) -> AxisSliceInfo {
+        AxisSliceInfo::NewAxis
+    }
+}
+
+/// A type that can slice an array of dimension `D`.
+///
+/// This trait is unsafe to implement because the implementation must ensure
+/// that `D`, `Self::OutDim`, `self.in_dim()`, and `self.out_ndim()` are
+/// consistent with the `&[AxisSliceInfo]` returned by `self.as_ref()`.
+pub unsafe trait CanSlice<D: Dimension>: AsRef<[AxisSliceInfo]> {
+    type OutDim: Dimension;
+
+    fn in_ndim(&self) -> usize;
+
+    fn out_ndim(&self) -> usize;
+}
+
+macro_rules! impl_canslice_samedim {
+    ($in_dim:ty) => {
+        unsafe impl<T, Do> CanSlice<$in_dim> for SliceInfo<T, $in_dim, Do>
+        where
+            T: AsRef<[AxisSliceInfo]>,
+            Do: Dimension,
+        {
+            type OutDim = Do;
+
+            fn in_ndim(&self) -> usize {
+                self.in_ndim()
+            }
+
+            fn out_ndim(&self) -> usize {
+                self.out_ndim()
+            }
+        }
+    };
+}
+impl_canslice_samedim!(Ix0);
+impl_canslice_samedim!(Ix1);
+impl_canslice_samedim!(Ix2);
+impl_canslice_samedim!(Ix3);
+impl_canslice_samedim!(Ix4);
+impl_canslice_samedim!(Ix5);
+impl_canslice_samedim!(Ix6);
+
+unsafe impl<T, Di, Do> CanSlice<IxDyn> for SliceInfo<T, Di, Do>
+where
+    T: AsRef<[AxisSliceInfo]>,
+    Di: Dimension,
+    Do: Dimension,
+{
+    type OutDim = Do;
+
+    fn in_ndim(&self) -> usize {
+        self.in_ndim()
+    }
+
+    fn out_ndim(&self) -> usize {
+        self.out_ndim()
+    }
+}
+
 /// Represents all of the necessary information to perform a slice.
 ///
-/// The type `T` is typically `[SliceOrIndex; n]`, `[SliceOrIndex]`, or
-/// `Vec<SliceOrIndex>`. The type `D` is the output dimension after calling
-/// [`.slice()`].
+/// The type `T` is typically `[AxisSliceInfo; n]`, `[AxisSliceInfo]`, or
+/// `Vec<AxisSliceInfo>`. The type `Di` is the dimension of the array to be
+/// sliced, and `Do` is the output dimension after calling [`.slice()`]. Note
+/// that if `Di` is a fixed dimension type (`Ix0`, `Ix1`, `Ix2`, etc.), the
+/// `SliceInfo` instance can still be used to slice an array with dimension
+/// `IxDyn` as long as the number of axes matches.
 ///
 /// [`.slice()`]: struct.ArrayBase.html#method.slice
 #[derive(Debug)]
 #[repr(C)]
-pub struct SliceInfo<T: ?Sized, D: Dimension> {
-    out_dim: PhantomData<D>,
+pub struct SliceInfo<T: ?Sized, Di: Dimension, Do: Dimension> {
+    in_dim: PhantomData<Di>,
+    out_dim: PhantomData<Do>,
     indices: T,
 }
 
-impl<T: ?Sized, D> Deref for SliceInfo<T, D>
+impl<T: ?Sized, Di, Do> Deref for SliceInfo<T, Di, Do>
```

Comment:
I suppose this could have been my idea in the start, but names like `Din` and `Dout` would help readability